### PR TITLE
Integer lists for Go

### PR DESCRIPTION
**Pull request by katrinwab**

### CODE REVIEW

**Comment by pascaldekloe** [comment]
Nice! Issue #26 is way overdue and version 2 might take a while.

Using an entire byte per element on the sign seems wasteful. Groups of 8 (flags) is a bit more complicated but doable. ZigZag-encoding makes the smallest size at the cost of speed. I think the performance loss might be worthwhile here. Whst do you think @katrinwab?

**Comment by katrinwab** [comment]
I changed the implementation to zigZag-encoding.

for an array of 1000 elements
BenchmarkMarshal-8                    10000            176800 ns/op
BenchmarkMarshalFloat-8            10000            173800 ns/op

for an array of 100000 elements
BenchmarkMarshal-8                     2000            755000 ns/op
BenchmarkMarshalFloat-8             3000            498666 ns/op

File: go.go [hunk 1]
```diff
@@ -259,6 +259,30 @@ const goMarshalField = `{{if eq .Type "bool"}}
...
+		x := uint(l)
+		for x >= 0x80 {
+			buf[i] = byte(x | 0x80)
+			x >>= 7
+			i++
+		}
+		buf[i] = byte(x)
+		i++
+		for _, v := range o.{{.NameTitle}} {
+			x1 := uint32((v << 1) ^ (v >> 31))
```
Comment (pascaldekloe) [inline go.go]: You can use the CPU's natural size as in `uint(v << 1) ^ uint(v >> 31)`.

File: go.go [hunk 2]
```diff
@@ -259,6 +259,30 @@ const goMarshalField = `{{if eq .Type "bool"}}
...
+			for n := 0; x1 >= 0x80 && n < 8; n++ {
```
Comment (pascaldekloe) [inline go.go]: The `n < 8` is a compiler hint. It should read 4 here though, with 32 bits.

Comment (pascaldekloe) [reply go.go]: Not even sure whether Go still needs it. The compiler has boundary knowledge these days. :-)

File: go.go [hunk 3]
```diff
@@ -446,6 +496,27 @@ const goMarshalFieldLen = `{{if eq .Type "bool"}}
...
+	if x := len(o.{{.NameTitle}}); x != 0 {
+		if x > ColferListMax {
+			return 0, ColferMax(fmt.Sprintf("colfer: field {{.String}} exceeds %d elements", ColferListMax))
+		}
+		for l += 2; x >= 0x80; l++ {
+			x >>= 7
+		}
+		for _, v := range o.{{.NameTitle}} {
+			x1 := uint32((v << 1) ^ (v >> 31))
+			l++
```
Comment (pascaldekloe) [inline go.go]: You can do this increment with one instruction outside of the loop as `l += len(o.{{.NameTitle}})`. Should be faster for large slices.

Comment (pascaldekloe) [reply go.go]: Dito for the 64-bit version below.

File: go.go (hunk 2 above)
Comment (pascaldekloe) [reply go.go]: Dito for the marshal length function below.

File: go.go (hunk 1 above)
Comment (pascaldekloe) [reply go.go]: Dito for the marshal length function below.

File: go.go [hunk 4]
```diff
@@ -687,6 +781,54 @@ const goUnmarshalField = `{{if eq .Type "bool"}}
...
+{{- if .TypeList}}
+	if header == {{.Index}} {
+	{{template "unmarshal-varint" .}}
+		if x > uint(ColferListMax) {
+			return 0, ColferMax(fmt.Sprintf("colfer: {{.String}} length %d exceeds %d elements", x, ColferListMax))
+		}
+
+		l := int(x)
+
+		if end := i + l*4; end >= len(data) {
```
Comment (pascaldekloe) [inline go.go]: Doesn't work here with variable length. This check can be removed, because the bounds are correctly validated in the loop below.

Comment (pascaldekloe) [reply go.go]: Dito for the 64-bit version below.

File: go.go [hunk 5]
```diff
@@ -687,6 +781,54 @@ const goUnmarshalField = `{{if eq .Type "bool"}}
...
+			goto eof
+		}
+		a := make([]int32, l)
+		for ai := range a {
+			if i+1 >= len(data) {
+				i++
+				goto eof
+			}
+
+			x := uint32(data[i])
```
Comment (pascaldekloe) [inline go.go]: Could use `uint` instead of `uint32` here and on line 810.

File: go.go [hunk 6]
```diff
@@ -687,6 +781,54 @@ const goUnmarshalField = `{{if eq .Type "bool"}}
...
+			if x >= 0x80 {
+				x &= 0x7f
+				for shift := uint(7); ; shift += 7 {
+					b := uint32(data[i])
+					i++
+					if i >= len(data) {
+						goto eof
+					}
+	
+					if b < 0x80 || shift == 56 {
```
Comment (pascaldekloe) [inline go.go]: The `shift == 56` is only for 64-bit integers (to prevent the 10-th byte worst-case).

File: schema.go [hunk 7]
```diff
@@ -94,7 +94,7 @@ func ParseFiles(files []string) ([]*Package, error) {
 				if ok {
 					if f.TypeList {
 						switch t {
-						case "float32", "float64", "text", "binary":
+						case "float32", "float64", "text", "binary", "int32", "int64":
```
Comment (pascaldekloe) [inline schema.go]: A compiler warning would be nice here, as this feature will fail miserably for other languages than Go.

```go
case "int32", "int64":
    log.Print("colfer: WARNING: integer lists are Go only at the moment")
```

File: go.go (hunk 1 above)
Comment (katrinwab) [reply go.go]: Then negative numbers take up more space
For example, -1690982682

File: go.go (hunk 2 above)
Comment (katrinwab) [reply go.go]: change this for 32 bits :)

File: go.go (hunk 3 above)
Comment (katrinwab) [reply go.go]: +

File: go.go (hunk 4 above)
Comment (katrinwab) [reply go.go]: changed

File: go.go (hunk 5 above)
Comment (katrinwab) [reply go.go]: changed

File: go.go (hunk 6 above)
Comment (katrinwab) [reply go.go]: fix it

File: schema.go (hunk 7 above)
Comment (katrinwab) [reply schema.go]: added )

**Comment by pascaldekloe** [comment]
Much appreciated @katrinwab 🙏☺️

**Pull request merged**